cpp/src/arrow/adapters/orc/adapter.cc

```diff
@@ -0,0 +1,699 @@
+    auto builder = static_cast<StructBuilder*>(abuilder);
+    auto batch = static_cast<liborc::StructVectorBatch*>(cbatch);
+
+    const uint8_t* valid_bytes = nullptr;
+    if (batch->hasNulls) {
+      valid_bytes = reinterpret_cast<const uint8_t*>(batch->notNull.data()) + offset;
+    }
+    RETURN_NOT_OK(builder->Append(length, valid_bytes));
+
+    int n_fields = builder->num_fields();
+    for (int i = 0; i < n_fields; i++) {
+      RETURN_NOT_OK(append_batch(type->getSubtype(i), builder->field_builder(i),
+                                 batch->fields[i], offset, length));
+    }
+    return Status::OK();
+  }
+
+  Status append_list_batch(const liborc::Type* type, ArrayBuilder* abuilder,
+                           liborc::ColumnVectorBatch* cbatch, int64_t offset,
+                           int64_t length) {
+    auto builder = static_cast<ListBuilder*>(abuilder);
+    auto batch = static_cast<liborc::ListVectorBatch*>(cbatch);
+    liborc::ColumnVectorBatch* elements = batch->elements.get();
+    const liborc::Type* elemtype = type->getSubtype(0);
+
+    for (int i = offset; i < length + offset; i++) {
+      if (!batch->hasNulls || batch->notNull[i]) {
+        int64_t start = batch->offsets[i];
+        int64_t end = batch->offsets[i + 1];
+        RETURN_NOT_OK(builder->Append());
+        RETURN_NOT_OK(append_batch(elemtype, builder->value_builder(), elements, start,
+                                   end - start));
+      } else {
+        RETURN_NOT_OK(builder->AppendNull());
+      }
+    }
+    return Status::OK();
+  }
+
+  Status append_map_batch(const liborc::Type* type, ArrayBuilder* abuilder,
+                          liborc::ColumnVectorBatch* cbatch, int64_t offset,
+                          int64_t length) {
+    auto list_builder = static_cast<ListBuilder*>(abuilder);
+    auto struct_builder = static_cast<StructBuilder*>(list_builder->value_builder());
+    auto batch = static_cast<liborc::MapVectorBatch*>(cbatch);
+    liborc::ColumnVectorBatch* keys = batch->keys.get();
+    liborc::ColumnVectorBatch* vals = batch->elements.get();
+    const liborc::Type* keytype = type->getSubtype(0);
+    const liborc::Type* valtype = type->getSubtype(1);
+
+    for (int i = offset; i < length + offset; i++) {
+      RETURN_NOT_OK(list_builder->Append());
+      int64_t start = batch->offsets[i];
+      int64_t list_length = batch->offsets[i + 1] - start;
+      if (list_length && (!batch->hasNulls || batch->notNull[i])) {
+        RETURN_NOT_OK(struct_builder->Append(list_length, nullptr));
+        RETURN_NOT_OK(append_batch(keytype, struct_builder->field_builder(0), keys, start,
+                                   list_length));
+        RETURN_NOT_OK(append_batch(valtype, struct_builder->field_builder(1), vals, start,
+                                   list_length));
+      }
+    }
+    return Status::OK();
+  }
+
+  template <class builder_type, class batch_type, class elem_type>
```

Comment:
Using `typename` instead of `class` might be more conforming, but I don't have a strong opinion